### ISSUE TITLE
Model format for exporting xgboost models to disk

### ISSUE THREAD

**Issue by akhildp** [dmlc/xgboost#5055]
https://github.com/dmlc/xgboost/blob/a4f5c862760029c24a5ba29b2a2ef4787058856c/python-package/xgboost/sklearn.py#L263

This warning here for `save_model` recommends using joblib for python and using the `save_model` only for a non-python binding of xgboost. If my use case was only restricted to using my model within different python environments only, is there any reason why joblib is recommended over xgboost's save and load other than the aforementioned reason of losing feature names?

I just want to understand the risks of joblib vs xgboost for deploying xgboost models to a different python environment to the one it was built in.

**Comment by trivialfis** [comment]
The feature names and types is saved in Python due to possible utf 8 string.  I think it's legacy and can be easily resolved by saving it in cpp, as I couldn't think of a reason why do we care about encodings.  We don't search nor query the encoding dependent string length.  @hcho3

**Comment by akhildp** [comment]
Other than that are there any other caveats of reading in a pickle file vs xgboost binary from a model performance/accuracy standpoint? Fort python, is there a reason for recommending to go with Joblib instead of xgboost binary? In other words does the format we choose impact the predictions at all? Sorry if this is a dumb question but I couldn't find anything in the documentation that states one way or the other.

**Comment by trivialfis** [comment]
@akhildp Would you like to take a look at the `saving_model.rst` in https://github.com/dmlc/xgboost/pull/4732 ?  Before that PR, there's no different other than `feature_names` and `feature_types`.  But after that PR there will be.

**Comment by trivialfis** [comment]
Also this one https://github.com/dmlc/xgboost/issues/4639

**Comment by akhildp** [comment]
Thanks for pointing this. These are really exciting changes. Is there an ETA for when these PRs may be merged and when 1.0.0 will be out?

**Comment by trivialfis** [comment]
My plan is to solve it first and then start bug squashing.  That PR fixes lots of issues especially for dask interface.  But it's also huge so I'm splitting it into smaller parts other can review.  You are more than welcomed to help the process.  There are already a few PRs I made with changes extracted from JSON PR.  Reviews are needed.
